src/bin/sccache-dist/main.rs

```diff
@@ -668,9 +675,77 @@ impl SchedulerIncoming for Scheduler {
         let mut servers = self.servers.lock().unwrap();
 
         match servers.get_mut(&server_id) {
-            Some(ref mut details) if details.server_nonce == server_nonce => {
-                details.last_seen = Instant::now();
-                return Ok(HeartbeatServerResult { is_new: false });
+            Some(ref mut details) => {
+                if details.server_nonce == server_nonce {
+                    let now = Instant::now();
+                    details.last_seen = now;
+
+                    let mut stale_jobs = Vec::new();
+                    for (&job_id, &last_seen) in details.jobs_unclaimed.iter() {
+                        if now.duration_since(last_seen) > UNCLAIMED_READY_TIMEOUT {
+                            let jobs = self.jobs.lock().unwrap();
+                            if let Some(detail) = jobs.get(&job_id) {
+                                match detail.state {
+                                    JobState::Ready => {
+                                        stale_jobs.push(job_id);
+                                    },
+                                    JobState::Pending => {
+                                        if now.duration_since(last_seen) >
+                                            UNCLAIMED_PENDING_TIMEOUT {
+                                                stale_jobs.push(job_id);
+                                            }
+                                    },
+                                    state => {
+                                        warn!("Invalid unclaimed job state for {}: {}",
+                                              job_id, state);
+                                    }
+                                }
+                            } else {
+                                warn!("Unknown stale job {}", job_id);
+                                stale_jobs.push(job_id);
+                            }
+                        }
+                    }
+
+                    if stale_jobs.len() > 0 {
+                        warn!("The following stale jobs will be de-allocated: {:?}",
+                              stale_jobs);
+
+                        let mut jobs = self.jobs.lock().unwrap();
+
+                        for job_id in stale_jobs {
+                            if !details.jobs_assigned.remove(&job_id) {
+                                warn!(
+                                    "Stale job for server {} not assigned: {}",
+                                    server_id.addr(), job_id
+                                );
+                            }
+                            if details.jobs_unclaimed.remove(&job_id).is_none() {
+                                warn!(
+                                    "Stale job for server {} not marked as unclaimed: {}",
```

Comment:
Yeah this is weird. It would mean it's not in the unclaimed list, which is how we got here in the first place. It "can't happen", this should probably become an assertion.